### PR TITLE
ANDAUTH-94 Integrate Github Actions

### PR DESCRIPTION
**Pull request by somenath1435**
Integrating Github Actions as Continuous Integration(CI) in the project.
This PR workflow has following tasks:

Cloning repo
Setting up JDK
Running unit tests
Building the project

### CODE REVIEW

**Comment by somenath1435** [comment]
@tmortagne @InsanusMokrassar please review, also running android tests is not preferred on ubuntu-latest image,  so a seprate job(running on mac-os) will be created later for that.

File: .github/workflows/build-test.yml [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+
+      - name: Set up JDK 1.8
+        uses: actions/setup-java@v1
+        with:
+          java-version: 1.8
+
+      - name: Run Unit tests
+        run: ./gradlew test --stacktrace
+
+      - name: Build project
```
Comment (InsanusMokrassar) [inline .github/workflows/build-test.yml:29]: I think, that building of project must be before the tests and fail the whole job in case of failure in build

**Review by InsanusMokrassar — CHANGES_REQUESTED** [review]

File: .github/workflows/build-test.yml (hunk 1 above)
Comment (somenath1435) [reply .github/workflows/build-test.yml:29]: Got it , I'll change

**Comment by InsanusMokrassar** [comment]
@somenath1435 Besides, I think, you could add config for labeler action in this pull request. Template for the config you may find [here (configuration of paths filters)](https://github.com/InsanusMokrassar/TelegramBotAPI/blob/master/.github/labeler.yml) and [here (config of labeler job)](https://github.com/InsanusMokrassar/TelegramBotAPI/blob/master/.github/workflows/label.yml)

**Comment by somenath1435** [comment]
@InsanusMokrassar adding labels as code/gradle/markdown as mentioned in the link should do the job or do I need to add more?

File: .github/workflows/build-test.yml [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+      - name: Clone Repo
+        uses: actions/checkout@v1
+
+      - name: Set up JDK 1.8
+        uses: actions/setup-java@v1
+        with:
+          java-version: 1.8
+
+      - name: Build project
+        run: ./gradlew build --continue
```
Comment (InsanusMokrassar) [inline .github/workflows/build-test.yml]: I think, it must be without `--continue`

**Review by InsanusMokrassar — CHANGES_REQUESTED** [review]
> adding labels as code/gradle/markdown as mentioned in the link should do the job or do I need to add more?

@somenath1435 For labeler I have added two files as mentioned in my [comment above](https://github.com/xwiki-contrib/android-authenticator/pull/64#issuecomment-637030223)

**Comment by somenath1435** [comment]
@InsanusMokrassar check now, suggest changes

**Review by InsanusMokrassar — COMMENTED** [review]
@somenath1435 Have you chacked it on some your own repo? Just to be sure that it is working

**Comment by tmortagne** [comment]
You sure this is enough ? Looks like a basic Java build without anything related to Android like setting Android SDK.

**Comment by somenath1435** [comment]
@tmortagne @InsanusMokrassar I have not tested the code on my repo, however similar code is working fine for other open source github android projects, I'm sharing links:

https://github.com/n8ebel/GitHubActionsAutomationSandbox/actions/runs/18348551/workflow
https://coletiv.com/blog/android-github-actions-setup/

Regarding setting up Android SDK, I haven't seen it anywhere till now.

**Comment by InsanusMokrassar** [comment]
> @tmortagne @InsanusMokrassar I have not tested the code on my repo, however similar code is working fine for other open source github android projects, I'm sharing links:
> 
> https://github.com/n8ebel/GitHubActionsAutomationSandbox/actions/runs/18348551/workflow
> https://coletiv.com/blog/android-github-actions-setup/
> 
> Regarding setting up Android SDK, I haven't seen it anywhere till now.

Anyway, I think, that it would be better to create an `Hello world` app and check that these configurations will correctly run at least unit tests

**Comment by somenath1435** [comment]
@InsanusMokrassar I've tested the workflow "Build-Test" on one of my repo, and it's working fine.
I think you have already tested the labeller workflow on your repo.

**Comment by tmortagne** [comment]
I find hard to believe that UI tests are working without configuring an emulator, are you sure they are executed ? I doubt github action always run one just on case, would be quite a waste of resources.

Another things which is certainly missing is the download and start/stop of an XWiki instance which is supposedly used by some UI tests (to not depends on a remote server which does not make any sense for automated tests).

You should probably take a look at https://ci.xwiki.org/view/All/job/Contrib_android-authenticator-docker/configure to see what is currently run during a Jenkins build.

**Comment by somenath1435** [comment]
@tmortagne I have mentioned that this workflow contains unit tests(which don't need emulator).
To do Android/UI/Instrumentation tests, we have 2 popular choices:
1. Run on mac-os using android emulator runner(preferred). [link](https://awesomeopensource.com/project/ReactiveCircus/android-emulator-runner)
2. Run on Firebase Test Lab.

I'm thinking of following 1st one and testing it on one of my repo. @InsanusMokrassar If you have better alternatives, please suggest.

**Comment by tmortagne** [comment]
> @tmortagne I have mentioned that this workflow contains unit tests(which don't need emulator).

OK but the point here is to build this project and this project have UI tests AFAIK.

**Comment by somenath1435** [comment]
@tmortagne I just ran the new workflow(running on mac-os) and it worked smoothly on one of my repo. I want to keep a seperate workflow dedicated to UI tests since it runs on mac-os virtual environment(HAMX already installed)

**Comment by somenath1435** [comment]
@InsanusMokrassar If you think, my approach is fine, I'll push the new workflow file

**Comment by somenath1435** [comment]
https://github.com/somenath1435/MathLib/actions/runs/123757358/workflow

**Comment by InsanusMokrassar** [comment]
I think, we could add one more option to run UI tests using some docker inside of action. You can find required configurations in our jenkins settings:)

**Comment by somenath1435** [comment]
@tmortagne @InsanusMokrassar I can't login to ci.xwiki.org, it's saying invalid username or password. I tried with my jenkins credentials. I can't see the config file for docker.

You can check the details of UI test job (in my repo) here:
https://github.com/somenath1435/MathLib/runs/734948380?check_suite_focus=true

**Comment by somenath1435** [comment]
@InsanusMokrassar Regarding UI test, what more option do you want to include? we can add more api levels. I found this link : https://github.com/ReactiveCircus/android-emulator-runner.
Note the current UI test boots up the emulator, do the tests and shuts it down. All happens in mac-os virtual environment.

**Comment by InsanusMokrassar** [comment]
@somenath1435 the point is to make it possible to test authenticator with UI tests during action

**Comment by somenath1435** [comment]
@InsanusMokrassar In my repo, UI tests are running fine. But do you have better alternatives?? If yes, please share links. Or do we need to write Dockerfile or docker scripts??

**Comment by InsanusMokrassar** [comment]
In current configuration I did not found any running of android emulator and tests on it. Can you, please, add variant which you have offered above and, I think, we will be able to try it

**Comment by somenath1435** [comment]
@InsanusMokrassar Please check my latest commit

**Comment by InsanusMokrassar** [comment]
As I understand, you have checked that connection to the device was established, but not run UI test. Have you checked that it?

**Comment by somenath1435** [comment]
@InsanusMokrassar I ran the same UI test workflow in one of my repo, and it not only connect to emulator but also run the UI tests(inside AndroidTest folder). Here is the details:
https://github.com/somenath1435/MathLib/runs/734948380?check_suite_focus=true

Am I missing something? Please look in the link and tell what I'm missing.

**Comment by InsanusMokrassar** [comment]
Can you, please, check that it is working by some special failing of UI test in your project?

**Comment by somenath1435** [comment]
@InsanusMokrassar The Ui Test is working as expected(failing where it should fail). I have tested it.

**Comment by InsanusMokrassar** [comment]
So, in this case I think we could try to merge it and look how it will work in our repo. What do you think @tmortagne ?

**Comment by tmortagne** [comment]
Sounds good yes, let's try.

**Review by tmortagne — APPROVED** [review]

**Review by InsanusMokrassar — APPROVED** [review]

File: .github/workflows/UI-Test.yml [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+    branches:
+      - 'master'
+  push:
+    branches:
+      - 'master'
+
+jobs:
+
+  UItest:
+    runs-on: macos-latest
```
Comment (vmassol) [inline .github/workflows/UI-Test.yml:14]: Why do you need to run this on macos (just curious)? Thanks

Comment (vmassol) [reply .github/workflows/UI-Test.yml:14]: General note: there's not many comments in this file and it's missing a license header

**Pull request merged**